### PR TITLE
Refactoring scikit_learn wrapper

### PR DESCRIPTION
**Pull request by ipod825**
Currently, the scikit_learn wrapper can not be used with scikit learn's grid_search api.
Also it's awkward to pass all the parameters used for 'fit', 'predict'... to the __init__ of the wrapper.

I attempt to solve this two problem by refactoring the wrapper. Details are documented in BaseWrapper. Use cases can be seen from the modified check_wrappers.py.

### CODE REVIEW

**Comment by fchollet** [comment]
The PR has merge conflicts with master which should be resolved first.

Please note that dosctrings are in Markdown and should follow the conventions found everywhere else in Keras.

**Comment by ipod825** [comment]
Is it possible to ```import sklearn``` in a testing file. I need to test and demonstrate that the wrapper could be used with the sklearn's grid_search API. But this will make the CI build failed.

**Comment by fchollet** [comment]
> Is it possible to import sklearn in a testing file?

No, let's not do that. Any other way you could test this? Can you not simply simulate the behavior of `grid_search` (the API calls it makes?). If not, we can leave it of the unit tests.

**Comment by ipod825** [comment]
I think currently I will just leave the `grid_search` test off. If I port things related to `grid_search` from `sklearn` into the testing script, things will be too messy.

I can think of one way to leave the testing script clean. That is, I port related `sklearn` functions in place like keras.utils.sk_utils.py. But I don't know if it's a good idea. 
The source directory of sklearn has an `externals` directory containing 3rd party utils (including joblib).
Is it possible to maintain 3rd party source somewhere in the same way?

File: keras/wrappers/scikit_learn.py [hunk 1]
```diff
@@ -75,7 +72,9 @@ def get_params(self, deep=True):
         params : dict
             Dictionary of parameter names mapped to their values.
         """
-        return {'model': self.model, 'optimizer': self.optimizer, 'loss': self.loss}
+        res = copy.deepcopy(self.sk_params)
+        res.update({'build_fn': self.build_fn})
```
Comment (fchollet) [inline keras/wrappers/scikit_learn.py]: why not `res['build_dn'] = self.build_fn`?

**Comment by fchollet** [comment]
Ok, it looks reasonable. Please:

- format all dosctrings in markdown, like those found everywhere else in Keras (e.g. look at `keras.layers.convolutional.py`).
- squash your commits into a single commit.

**Comment by ipod825** [comment]
Uh, I am not very familiar with git. I didn't create a branch at my clone repository for modification. I seems that if I use
 ```bash
git reset --hard <commit before my first commit>
git merge --squash HEAD@{1}
git push -f origin master
``` 
I will include commits done by others in this single commit.
Is that ok?

**Comment by fchollet** [comment]
You should first rebase your local repo to master. You can find
instructions online, also there's a link in our CONTRIBUTING.md file.

**Comment by ipod825** [comment]
I did not create another branch. So what I did is to reset my HEAD to the commit before my first commit and then rebase to master. Anyway, I've already squashed the commits. If there's a problem, please let me know.

**Comment by fchollet** [comment]
It's a mess. Please fix it...

**Comment by ipod825** [comment]
I don't know how to solve the problem. So I created another pull request #1842

**Comment by ipod825** [comment]
Squashed... CI build still failed

**Pull request closed without merging**